File: TREE/Python/LC 144 - BINARY TREE PREORDER TRAVERSAL.py
# LEETCODE 144 - BINARY TREE PREORDER TRAVERSAL
# QUESTION LINK - "https://leetcode.com/problems/binary-tree-preorder-traversal/"

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def preOrder(self, root, res):
        if not root:
            return
        res.append(root.val)
        self.preOrder(root.left, res)
        self.preOrder(root.right, res)

    def preorderTraversal(self, root):
        res = []
        self.preOrder(root, res)
        return res

File: TREE/Java/LC 144 - BINARY TREE PREORDER TRAVERSAL.java
// LEETCODE 144 - BINARY TREE PREORDER TRAVERSAL
// QUESTION LINK - "https://leetcode.com/problems/binary-tree-preorder-traversal/"

import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    private void preOrder(TreeNode root, List<Integer> res) {
        if (root == null) return;
        res.add(root.val);
        preOrder(root.left, res);
        preOrder(root.right, res);
    }

    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        preOrder(root, res);
        return res;
    }
}
